File: Question 0079 - Word Search/src/Solution.java
class Solution {
    private boolean dfs(String word, char[][] board, int r, int c)
    {
        if(word.isEmpty()) return true;

        if(r<0 || c<0 || r>= board.length || c>=board[0].length || board[r][c] != word.charAt(0)) return false;


        char temp = board[r][c];
        board[r][c] = '#';
        boolean up = dfs(word.substring(1), board, r+1, c);
        boolean down = dfs(word.substring(1), board, r-1, c);
        boolean right = dfs(word.substring(1), board, r, c+1);
        boolean left = dfs(word.substring(1), board, r, c-1);
        board[r][c] = temp;
        return up || down || right||left;

    }
    public boolean exist(char[][] board, String word)
    {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if(dfs(word,board,i,j)) return true;
            }
        }

        return false;
    }
}

File: Question 0079 - Word Search/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

bool solve(vector<vector<char>>& board, string& word, int i, int j, vector<vector<int>>& dir, int idx)
{
    if(board[i][j] != word[idx]) return false;
    if(idx == word.size() - 1) return true;
    char temp = board[i][j];
    board[i][j] = '#';

    for(auto& d : dir)
    {
        int x = i + d[0];
        int y = j + d[1];

        if(x >= 0 && x < board.size() && y >= 0 && y < board[0].size())
        {
            if (solve(board, word, x, y, dir, idx+1)) return true;
        }
    }

    board[i][j] = temp;
    return false;
}

class Solution 
{
public:
    bool exist(vector<vector<char>>& board, string word) 
    {
        int r = board.size(), c = board[0].size();
        vector<vector<int>> dir = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        for(int i = 0; i < r; i++)
        {
            for(int j = 0; j < c; j++)
            {
                if(solve(board, word, i, j, dir, 0)) return true;
            }
        }

        return false;
    }
};